### DISCUSSION TITLE
On using exceptions for normal control flow

### DISCUSSION THREAD

**Discussion by verdverm** [vikejs/vike#1517] [Discussion]
I'm curious about the decision to use exception as a way to control normal logic flow. This is normally considered an anti-pattern (https://www.google.com/search?q=using+exceptions+for+control+flow)

https://vike.dev/redirect

```ts
import { redirect } from 'vike/abort'

function onSomeHook() {
  if (someCondition) {
    throw redirect('/some-url')
  }
}
```

Is this pattern something you expect to remove in time?

**Comment by brillout** [comment]
I agree in principle. That said, and I stand to be corrected, for this particular use case ~~I don't see a concrete issue~~ (**edit**: see my reply down below). (While the upside is that you can throw everywhere even in a UI component if your UI framewrok passes through the error to `onRender{Client,Html}()`.)

**Reply by verdverm** [reply]
I'm no expert on this, a few points I'm aware of are
1. exceptions generally come with significant overhead compared to function calls
2. the other frameworks don't have this pattern, afiak
3. regardless of the merits, people will see this and have concern for code quality

**Reply by verdverm** [reply]
A contrived example

```ts
function onSomeHook() {
  try {
    // this package can throw exceptions
    const ret = call3rdPartyLibrary();
    if (ret.redirect) {
      throw redirect(ret.redirect)
    }
  } catch (e) {
    // handle error?
    console.error(e)

    // throw redirect to error page?
    // how to pass the actual error as context?
    throw redirect("/error-page")
  }
}
```

**Reply by phonzammi** [reply]
Hi @verdverm 
Sorry to interrupt

> throw redirect to error page?
> how to pass the actual error as context?


Have you ever try to just `throw error` ?
It will automatically render an error page, then check in pageContext for the error context

**Reply by brillout** [reply]
One thing I do dislike about using exceptions is that it breaks the explicitness of the logical flow. Having to dig into why something's happening because the logic isn't collocated is a considerable DX degradation.

That said, I do feel like there is a valid use case for using exceptions (and that's why my intuition led me down that path). Most notably when using RPC.

When using RPC (e.g. [Telefunc](https://telefunc.com/)), how can the user redirect/rewrite depending on the result of the fetched data? E.g. to redirect the user to `/wiki/add/movie` if the user goes to `/wiki/movie/star-wars-12` but the wiki page for the movie with the slug `star-wars-12` wasn't created yet. The issue here is that telefunctions aren't Vike hooks.

If we do find a solution for RPC, then we could do that: we support `return { redirect, render }` to `guard()` and add a deprecation warning saying "please use `guard()` instead of `throw render()` / `throw redirect()`". We'll then see whether users will complain about that and miss being able to use exceptions. (Depending on what users say we either keep exceptions and revert the deprecation warning, or remove the exceptions.)

We'll know a lot more after we further work on the deep Vike + Telefunc integration (there is a working prototype but isn't finished yet).

(As for the other reasons you listed, I think they're negligible.)
